Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs

```diff
@@ -285,6 +283,71 @@ public void TestGetCurrentPoolSizeReturnsSumOfPoolSizes()
             Assert.AreEqual(5, poolSize);
         }
 
+        [Test]
+        public void TestReturnPoolForSecurePassword()
+        {
+            // arrange
+            const string AnotherPassword = "anotherPassword";
+            EnsurePoolSize(ConnectionStringWithoutPassword, _password3, 1);
+
+            // act
+            var pool = _connectionPoolManager.GetPool(ConnectionStringWithoutPassword, SecureStringHelper.Encode(AnotherPassword)); // a new pool has been created because the password is different
+
+            // assert
+            Assert.AreEqual(0, pool.GetCurrentPoolSize());
+            Assert.AreEqual(AnotherPassword, SecureStringHelper.Decode(pool.Password));
+        }
+
+        [Test]
+        public void TestReturnDifferentPoolWhenPasswordProvidedInDifferentWay()
+        {
+            // arrange
+            var connectionStringWithPassword = $"{ConnectionStringWithoutPassword}password={SecureStringHelper.Decode(_password3)}";
+            EnsurePoolSize(ConnectionStringWithoutPassword, _password3, 2);
+            EnsurePoolSize(connectionStringWithPassword, null, 5);
+            EnsurePoolSize(connectionStringWithPassword, _password3, 8);
+
+            // act
+            var pool1 = _connectionPoolManager.GetPool(ConnectionStringWithoutPassword, _password3);
+            var pool2 = _connectionPoolManager.GetPool(connectionStringWithPassword, null);
+            var pool3 = _connectionPoolManager.GetPool(connectionStringWithPassword, _password3);
+
+            // assert
+            Assert.AreEqual(2, pool1.GetCurrentPoolSize());
+            Assert.AreEqual(5, pool2.GetCurrentPoolSize());
+            Assert.AreEqual(8, pool3.GetCurrentPoolSize());
+        }
+
+        [Test]
+        [TestCase(null)]
+        [TestCase("")]
+        public void TestGetPoolFailsWhenNoPasswordProvided(string password)
+        {
+            // arrange
+            var securePassword = password == null ? null : SecureStringHelper.Encode(password);
+
+            // act
+            var thrown = Assert.Throws<SnowflakeDbException>(() => _connectionPoolManager.GetPool(ConnectionStringWithoutPassword, securePassword));
+
+            // assert
+            Assert.That(thrown.Message, Does.Contain("Required property PASSWORD is not provided"));
+        }
+
+        [Test]
+        public void TestPoolDoesNotSerializePassword()
+        {
+            // arrange
+            var password = SecureStringHelper.Decode(_password3);
+            var connectionStringWithPassword = $"{ConnectionStringWithoutPassword}password={password}";
+            var pool = _connectionPoolManager.GetPool(connectionStringWithPassword, _password3);
+
+            // act
+            var serializedPool = pool.ToString();
+
+            // assert
+            Assert.IsFalse(serializedPool.Contains(password));
+        }
```

Comment:
should there be TestThrowsWhenPasswordNotGivenInConnectionStringNorSecurePassword ?

Reply:
SessionPoolTest has such a test, but actually this validation is a kind of a double check. Normally in the pool manager if you search for the pool with wrong password the pool with the wrong password is created and returned.

Reply:
added also tests for password required